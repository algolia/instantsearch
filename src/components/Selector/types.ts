export type SelectorProps = {
  cssClasses?: {
    root?: string | string[];
    select?: string | string[];
    option?: string | string[];
  };
  currentValue?: string | number;
  options: Array<{
    value?: string | number;
    label: string;
  }>;
  setValue(value: string): void;
};
