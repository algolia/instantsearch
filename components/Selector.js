let React = require('react');

class Selector extends React.Component {
  handleChange(event) {
    this.props.setValue(event.target.value);
  }

  render() {
    let {currentValue, options} = this.props;

    let handleChange = this.handleChange.bind(this);

    return (
      <select
        className={this.props.cssClasses.root}
        onChange={handleChange}
        value={currentValue}
      >
        {options.map((option) => {
          return <option className={this.props.cssClasses.item} key={option.value} value={option.value}>{option.label}</option>;
        })}
      </select>
    );
  }
}

Selector.propTypes = {
  cssClasses: React.PropTypes.shape({
    root: React.PropTypes.oneOfType([
      React.PropTypes.string,
      React.PropTypes.arrayOf(React.PropTypes.string)
    ]),
    item: React.PropTypes.oneOfType([
      React.PropTypes.string,
      React.PropTypes.arrayOf(React.PropTypes.string)
    ])
  }),
  currentValue: React.PropTypes.oneOfType([
    React.PropTypes.string,
    React.PropTypes.number
  ]).isRequired,
  options: React.PropTypes.arrayOf(
    React.PropTypes.shape({
      value: React.PropTypes.oneOfType([
        React.PropTypes.string,
        React.PropTypes.number
      ]).isRequired,
      label: React.PropTypes.string.isRequired
    })
  ).isRequired,
  setValue: React.PropTypes.func.isRequired
};

module.exports = Selector;
