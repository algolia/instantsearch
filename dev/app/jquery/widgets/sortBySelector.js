/* eslint-disable import/default */
import instantsearch from '../../../../index';

const renderFn = (
  { options, refine, currentRefinement, widgetParams: { containerNode } },
  isFirstRendering
) => {
  if (isFirstRendering) {
    containerNode.append('<select></select>');
    containerNode
      .find('select')
      .on('change', ({ target: { value } }) => refine(value));
  }

  const optionsHTML = options.map(
    ({ label, value }) => `
    <option
      value="${value}"
      ${currentRefinement === value ? 'selected' : ''}
    >
      ${label}
    </option>
  `
  );

  containerNode.find('select').html(optionsHTML);
};

export default instantsearch.connectors.connectSortBySelector(renderFn);
