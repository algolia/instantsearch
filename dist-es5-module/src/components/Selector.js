'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var Selector = function (_React$Component) {
  _inherits(Selector, _React$Component);

  function Selector() {
    _classCallCheck(this, Selector);

    return _possibleConstructorReturn(this, (Selector.__proto__ || Object.getPrototypeOf(Selector)).apply(this, arguments));
  }

  _createClass(Selector, [{
    key: 'componentWillMount',
    value: function componentWillMount() {
      this.handleChange = this.handleChange.bind(this);
    }
  }, {
    key: 'handleChange',
    value: function handleChange(event) {
      this.props.setValue(event.target.value);
    }
  }, {
    key: 'render',
    value: function render() {
      var _this2 = this;

      var _props = this.props;
      var currentValue = _props.currentValue;
      var options = _props.options;


      return _react2.default.createElement(
        'select',
        {
          className: this.props.cssClasses.root,
          onChange: this.handleChange,
          value: currentValue
        },
        options.map(function (option) {
          return _react2.default.createElement(
            'option',
            {
              className: _this2.props.cssClasses.item,
              key: option.value,
              value: option.value },
            option.label
          );
        })
      );
    }
  }]);

  return Selector;
}(_react2.default.Component);

Selector.propTypes = {
  cssClasses: _react2.default.PropTypes.shape({
    root: _react2.default.PropTypes.oneOfType([_react2.default.PropTypes.string, _react2.default.PropTypes.arrayOf(_react2.default.PropTypes.string)]),
    item: _react2.default.PropTypes.oneOfType([_react2.default.PropTypes.string, _react2.default.PropTypes.arrayOf(_react2.default.PropTypes.string)])
  }),
  currentValue: _react2.default.PropTypes.oneOfType([_react2.default.PropTypes.string, _react2.default.PropTypes.number]).isRequired,
  options: _react2.default.PropTypes.arrayOf(_react2.default.PropTypes.shape({
    value: _react2.default.PropTypes.oneOfType([_react2.default.PropTypes.string, _react2.default.PropTypes.number]).isRequired,
    label: _react2.default.PropTypes.string.isRequired
  })).isRequired,
  setValue: _react2.default.PropTypes.func.isRequired
};

exports.default = Selector;