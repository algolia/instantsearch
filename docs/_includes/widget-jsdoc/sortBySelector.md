<h4 class="no-toc">Parameters</h4>
<p class="attr-name">
<span class='attr-required'>`options.container`<span class="show-description">…</span></span>
  <span class="attr-infos">(<code>string</code> &#124; <code>DOMElement</code>)</span>
</p>
<p class="attr-description important">CSS Selector or DOMElement to insert the widget</p>
<p class="attr-name">
<span class='attr-required'>`options.indices`<span class="show-description">…</span></span>
  <span class="attr-infos">(<code>Array</code>)</span>
</p>
<p class="attr-description important">Array of objects defining the different indices to choose from.</p>
<p class="attr-name">
<span class='attr-required'>`options.indices[0].name`<span class="show-description">…</span></span>
  <span class="attr-infos">(<code>string</code>)</span>
</p>
<p class="attr-description important">Name of the index to target</p>
<p class="attr-name">
<span class='attr-required'>`options.indices[0].label`<span class="show-description">…</span></span>
  <span class="attr-infos">(<code>string</code>)</span>
</p>
<p class="attr-description important">Label displayed in the dropdown</p>
<p class="attr-name">
<span class='attr-optional'>`options.autoHideContainer`<span class="show-description">…</span></span>
  <span class="attr-infos">Default:<code class="attr-default">false</code>(<code>boolean</code>)</span>
</p>
<p class="attr-description">Hide the container when no results match</p>
<p class="attr-name">
<span class='attr-optional'>`options.cssClasses`<span class="show-description">…</span></span>
  <span class="attr-infos">(<code>Object</code>)</span>
</p>
<p class="attr-description">CSS classes to be added</p>
<p class="attr-name">
<span class='attr-optional'>`options.cssClasses.root`<span class="show-description">…</span></span>
  <span class="attr-infos">(<code>string</code> &#124; <code>Array.&lt;string&gt;</code>)</span>
</p>
<p class="attr-description">CSS classes added to the parent <select></p>
<p class="attr-name">
<span class='attr-optional'>`options.cssClasses.item`<span class="show-description">…</span></span>
  <span class="attr-infos">(<code>string</code> &#124; <code>Array.&lt;string&gt;</code>)</span>
</p>
<p class="attr-description">CSS classes added to each <option></p>

<p class="attr-legend">* <span>Required</span></p>
