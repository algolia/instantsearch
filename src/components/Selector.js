import React from 'react';

class Selector extends React.Component {
  componentWillMount() {
    this.handleChange = this.handleChange.bind(this);
  }

  shouldComponentUpdate() {
    return false;
  }

  handleChange(event) {
    this.props.setValue(event.target.value);
  }

  render() {
    const {currentValue, options} = this.props;

    return (
      <select
        className={this.props.cssClasses.root}
        defaultValue={currentValue}
        onChange={this.handleChange}
      >
        {options.map(option =>
          <option
            className={this.props.cssClasses.item}
            key={option.value}
            value={option.value}>
              {option.label}
          </option>
        )}
      </select>
    );
  }
}

Selector.propTypes = {
  cssClasses: React.PropTypes.shape({
    root: React.PropTypes.oneOfType([
      React.PropTypes.string,
      React.PropTypes.arrayOf(React.PropTypes.string)
    ]),
    item: React.PropTypes.oneOfType([
      React.PropTypes.string,
      React.PropTypes.arrayOf(React.PropTypes.string)
    ])
  }),
  currentValue: React.PropTypes.oneOfType([
    React.PropTypes.string,
    React.PropTypes.number
  ]).isRequired,
  options: React.PropTypes.arrayOf(
    React.PropTypes.shape({
      value: React.PropTypes.oneOfType([
        React.PropTypes.string,
        React.PropTypes.number
      ]).isRequired,
      label: React.PropTypes.string.isRequired
    })
  ).isRequired,
  setValue: React.PropTypes.func.isRequired
};

export default Selector;
