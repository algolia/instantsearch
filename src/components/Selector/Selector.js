/** @jsx h */

import { h } from 'preact';
import PropTypes from 'prop-types';
import cx from 'classnames';

function Selector({ currentValue, options, cssClasses, setValue }) {
  return (
    <select
      className={cx(cssClasses.select)}
      onChange={event => setValue(event.target.value)}
      value={`${currentValue}`}
    >
      {options.map(option => (
        <option
          className={cx(cssClasses.option)}
          key={option.label + option.value}
          value={`${option.value}`}
        >
          {option.label}
        </option>
      ))}
    </select>
  );
}

Selector.propTypes = {
  cssClasses: PropTypes.shape({
    root: PropTypes.oneOfType([
      PropTypes.string,
      PropTypes.arrayOf(PropTypes.string),
    ]),
    select: PropTypes.oneOfType([
      PropTypes.string,
      PropTypes.arrayOf(PropTypes.string),
    ]),
    option: PropTypes.oneOfType([
      PropTypes.string,
      PropTypes.arrayOf(PropTypes.string),
    ]),
  }),
  currentValue: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  options: PropTypes.arrayOf(
    PropTypes.shape({
      value: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
      label: PropTypes.string.isRequired,
    })
  ).isRequired,
  setValue: PropTypes.func.isRequired,
};

export default Selector;
