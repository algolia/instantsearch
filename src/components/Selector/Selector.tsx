/** @jsx h */

import { h } from 'preact';
import cx from 'classnames';

export type SelectorOption = {
  value?: string | number;
  label: string;
};

export type SelectorComponentCSSClasses = {
  root: string;
  select: string;
  option: string;
};

export type SelectorProps = {
  cssClasses: SelectorComponentCSSClasses;
  currentValue?: string | number;
  options: SelectorOption[];
  setValue(value: SelectorOption['value']): void;
};

function Selector({
  currentValue,
  options,
  cssClasses,
  setValue,
}: SelectorProps) {
  return (
    <select
      className={cx(cssClasses.select)}
      onChange={(event) => setValue((event.target as HTMLSelectElement).value)}
      value={`${currentValue}`}
    >
      {options.map((option) => (
        <option
          className={cx(cssClasses.option)}
          key={option.label + option.value}
          value={`${option.value}`}
        >
          {option.label}
        </option>
      ))}
    </select>
  );
}

export default Selector;
