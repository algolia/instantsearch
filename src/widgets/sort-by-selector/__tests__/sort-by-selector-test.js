import sinon from 'sinon';
import expect from 'expect';
import sortBySelector from '../sort-by-selector';
import Selector from '../../../components/Selector';

import instantSearch from '../../../lib/main';

describe('sortBySelector call', () => {
  it('throws an exception when no options', () => {
    const container = document.createElement('div');
    expect(sortBySelector.bind(null, { container })).toThrow(/^Usage/);
  });

  it('throws an exception when no indices', () => {
    const indices = [];
    expect(sortBySelector.bind(null, { indices })).toThrow(/^Usage/);
  });
});

describe('sortBySelector()', () => {
  let ReactDOM;
  let container;
  let indices;
  let cssClasses;
  let widget;
  let helper;
  let results;
  let autoHideContainer;

  beforeEach(() => {
    const instantSearchInstance = instantSearch({
      apiKey: '',
      appId: '',
      indexName: 'defaultIndex',
      createAlgoliaClient: () => ({}),
    });
    autoHideContainer = sinon.stub().returns(Selector);
    ReactDOM = { render: sinon.spy() };

    sortBySelector.__Rewire__('render', ReactDOM.render);
    sortBySelector.__Rewire__('autoHideContainerHOC', autoHideContainer);

    container = document.createElement('div');
    indices = [
      { name: 'index-a', label: 'Index A' },
      { name: 'index-b', label: 'Index B' },
    ];
    cssClasses = {
      root: ['custom-root', 'cx'],
      select: 'custom-select',
      item: 'custom-item',
    };
    widget = sortBySelector({ container, indices, cssClasses });
    helper = {
      getIndex: sinon.stub().returns('index-a'),
      setIndex: sinon.stub().returnsThis(),
      search: sinon.spy(),
    };

    results = {
      hits: [],
      nbHits: 0,
    };
    widget.init({ helper, instantSearchInstance });
  });

  it("doesn't configure anything", () => {
    expect(widget.getConfiguration).toEqual(undefined);
  });

  it('calls twice ReactDOM.render(<Selector props />, container)', () => {
    widget.render({ helper, results });
    widget.render({ helper, results });
    expect(ReactDOM.render.calledTwice).toBe(
      true,
      'ReactDOM.render called twice'
    );
    expect(ReactDOM.render.firstCall.args[0]).toMatchSnapshot();
    expect(ReactDOM.render.firstCall.args[1]).toEqual(container);
    expect(ReactDOM.render.secondCall.args[0]).toMatchSnapshot();
    expect(ReactDOM.render.secondCall.args[1]).toEqual(container);
  });

  it('sets the underlying index', () => {
    widget.setIndex('index-b');
    expect(helper.setIndex.calledOnce).toBe(true, 'setIndex called once');
    expect(helper.search.calledOnce).toBe(true, 'search called once');
  });

  it('should throw if there is no name attribute in a passed object', () => {
    indices.length = 0;
    indices.push({ label: 'Label without a name' });
    expect(() => {
      widget.init({ helper });
    }).toThrow(/Index index-a not present/);
  });

  it('must include the current index at initialization time', () => {
    helper.getIndex = sinon.stub().returns('non-existing-index');
    expect(() => {
      widget.init({ helper });
    }).toThrow(/Index non-existing-index not present/);
  });

  afterEach(() => {
    sortBySelector.__ResetDependency__('render');
    sortBySelector.__ResetDependency__('autoHideContainerHOC');
  });
});
