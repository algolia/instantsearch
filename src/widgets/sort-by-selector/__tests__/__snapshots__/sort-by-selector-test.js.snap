// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`sortBySelector() calls twice ReactDOM.render(<Selector props />, container) 1`] = `
<HeaderFooter-AutoHide
  cssClasses={
    Object {
      "item": "ais-sort-by-selector--item custom-item",
      "root": "ais-sort-by-selector custom-root cx",
      "select": "ais-sort-by-selector custom-select",
    }
  }
  currentValue="index-a"
  options={
    Array [
      Object {
        "label": "Index A",
        "value": "index-a",
      },
      Object {
        "label": "Index B",
        "value": "index-b",
      },
    ]
  }
  setValue={[Function]}
  shouldAutoHideContainer={false}
/>
`;

exports[`sortBySelector() calls twice ReactDOM.render(<Selector props />, container) 2`] = `
<HeaderFooter-AutoHide
  cssClasses={
    Object {
      "item": "ais-sort-by-selector--item custom-item",
      "root": "ais-sort-by-selector custom-root cx",
      "select": "ais-sort-by-selector custom-select",
    }
  }
  currentValue="index-a"
  options={
    Array [
      Object {
        "label": "Index A",
        "value": "index-a",
      },
      Object {
        "label": "Index B",
        "value": "index-b",
      },
    ]
  }
  setValue={[Function]}
  shouldAutoHideContainer={false}
/>
`;

exports[`sortBySelector() renders transformed items 1`] = `
<HeaderFooter-AutoHide
  cssClasses={
    Object {
      "item": "ais-sort-by-selector--item",
      "root": "ais-sort-by-selector",
      "select": "ais-sort-by-selector",
    }
  }
  currentValue="index-a"
  options={
    Array [
      Object {
        "label": "Index A",
        "transformed": true,
        "value": "index-a",
      },
      Object {
        "label": "Index B",
        "transformed": true,
        "value": "index-b",
      },
    ]
  }
  setValue={[Function]}
  shouldAutoHideContainer={false}
/>
`;
